File: projects/backend/services/auth-service/src/auth_service/services/audit.py
"""Audit service."""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from auth_service.domain.models import AuditEntry
from auth_service.repositories.audit import AuditRepository


class AuditService:
    """Service for recording and querying the audit log."""

    def __init__(self, audit_repo: AuditRepository) -> None:
        self._repo = audit_repo

    async def log_action(
        self,
        actor_id: UUID,
        action: str,
        scope_type: str,
        *,
        scope_id: UUID | None = None,
        target_type: str | None = None,
        target_id: str | None = None,
        details: dict[str, Any] | None = None,
        ip_address: str | None = None,
        user_agent: str | None = None,
    ) -> AuditEntry:
        """Record an audit log entry."""
        return await self._repo.log(
            actor_id=actor_id,
            action=action,
            scope_type=scope_type,
            scope_id=scope_id,
            target_type=target_type,
            target_id=target_id,
            details=details,
            ip_address=ip_address,
            user_agent=user_agent,
        )

    async def query(
        self,
        *,
        actor_id: UUID | None = None,
        action: str | None = None,
        scope_type: str | None = None,
        scope_id: UUID | None = None,
        target_type: str | None = None,
        target_id: str | None = None,
        from_date: datetime | None = None,
        to_date: datetime | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[AuditEntry]:
        """Query audit log with filters and pagination."""
        return await self._repo.query(
            actor_id=actor_id,
            action=action,
            scope_type=scope_type,
            scope_id=scope_id,
            target_type=target_type,
            target_id=target_id,
            from_date=from_date,
            to_date=to_date,
            limit=limit,
            offset=offset,
        )

File: projects/frontend/apps/experiment-portal/src/api/audit.ts
import type { AuditLogResponse } from '../types/permissions'
import { apiGet } from './client'

export interface AuditLogFilters {
  actor_id?: string
  action?: string
  scope_type?: 'system' | 'project'
  scope_id?: string
  target_type?: string
  target_id?: string
  from?: string
  to?: string
  limit?: number
  offset?: number
}

export const auditApi = {
  queryAuditLog: async (filters: AuditLogFilters = {}): Promise<AuditLogResponse> => {
    const params: Record<string, string | number> = {}
    if (filters.actor_id) params.actor_id = filters.actor_id
    if (filters.action) params.action = filters.action
    if (filters.scope_type) params.scope_type = filters.scope_type
    if (filters.scope_id) params.scope_id = filters.scope_id
    if (filters.target_type) params.target_type = filters.target_type
    if (filters.target_id) params.target_id = filters.target_id
    if (filters.from) params.from = filters.from
    if (filters.to) params.to = filters.to
    if (filters.limit !== undefined) params.limit = filters.limit
    if (filters.offset !== undefined) params.offset = filters.offset
    return apiGet<AuditLogResponse>('/api/v1/audit-log', { params })
  },
}
